# Modificando dados com Pandas - parte 2

## Adicionando colunas

Inicialmente vamos deixar o DataFrame com as modificações da parte 1.

In [2]:
import pandas as pd # type: ignore
dados_ibov = pd.read_csv('^BVSP.csv')
# alterando os nomes das colunas e o índice (ver parte 1)
dados_ibov.columns = dados_ibov.columns.str.replace(' ', '_').str.lower()
dados_ibov.set_index('date', inplace=True)
dados_ibov.head(10)

open      high       low     close  adj_close    volume
date                                                                   
2023-08-03  120859.0  122619.0  120365.0  120586.0   120586.0  13027300
2023-08-04  120586.0  121442.0  119215.0  119508.0   119508.0  15065200
2023-08-07  119509.0  120104.0  118894.0  119380.0   119380.0   8726500
2023-08-08  119378.0  119553.0  117492.0  119090.0   119090.0  10246000
2023-08-09  119090.0  119090.0  117901.0  118409.0   118409.0  11245600
2023-08-10  118412.0  119438.0  118113.0  118350.0   118350.0  12689600
2023-08-11  118350.0  119054.0  117415.0  118065.0   118065.0  11867500
2023-08-14  118067.0  118082.0  116530.0  116810.0   116810.0  11203500
2023-08-15  116809.0  117697.0  116033.0  116171.0   116171.0  14636000
2023-08-16  116171.0  117338.0  115534.0  115592.0   115592.0  18862600

Nesse exemplo vamos adicionar uma coluna chamada `return_close` com o valores 0 ao DataFrame.

In [3]:
import numpy as np
dados_ibov['return_close'] = np.zeros(dados_ibov.shape[0])
dados_ibov.head(10)

open      high       low     close  adj_close    volume  \
date                                                                      
2023-08-03  120859.0  122619.0  120365.0  120586.0   120586.0  13027300   
2023-08-04  120586.0  121442.0  119215.0  119508.0   119508.0  15065200   
2023-08-07  119509.0  120104.0  118894.0  119380.0   119380.0   8726500   
2023-08-08  119378.0  119553.0  117492.0  119090.0   119090.0  10246000   
2023-08-09  119090.0  119090.0  117901.0  118409.0   118409.0  11245600   
2023-08-10  118412.0  119438.0  118113.0  118350.0   118350.0  12689600   
2023-08-11  118350.0  119054.0  117415.0  118065.0   118065.0  11867500   
2023-08-14  118067.0  118082.0  116530.0  116810.0   116810.0  11203500   
2023-08-15  116809.0  117697.0  116033.0  116171.0   116171.0  14636000   
2023-08-16  116171.0  117338.0  115534.0  115592.0   115592.0  18862600   

            return_close  
date                      
2023-08-03           0.0  
2023-08-04           0.0  
2023-08-07           0.0  
2023-08-08           0.0  
2023-08-09           0.0  
2023-08-10           0.0  
2023-08-11           0.0  
2023-08-14           0.0  
2023-08-15           0.0  
2023-08-16           0.0

A função `zeros` do NumPy retorna um array com zeros do tamanho especificado. No nosso caso, o número de elemntos do array deve ser exatamente igual ao numero de linhas no DataFrame. Para obter o número de linhas, a propriedade `shape` da classe DataFrame é útil, pois ela retorna uma tupla com as dimensões do objeto:

In [5]:
dados_ibov.shape

(250, 7)

Isso indica que nosso DataFrame tem 250 linhas e 7 colunas (já contando a nova).

Entretanto, uma nova coluna apenas com zeros não ajuda muito. Para calcular os retornos, devemos calcular a variação percentual de um dia para o outro no valor de fechamento (ou no fechamento ajustado, conforme necessidade). A biblioteca Pandas já possui uma função pronta para isso, chamada `pct_change()`.

In [21]:
dados_ibov.close.pct_change()

date
2023-08-03         NaN
2023-08-04   -0.008940
2023-08-07   -0.001071
2023-08-08   -0.002429
2023-08-09   -0.005718
                ...   
2024-07-29   -0.004220
2024-07-30   -0.006420
2024-07-31    0.011995
2024-08-01   -0.002013
2024-08-02   -0.012096
Name: close, Length: 250, dtype: float64

Note o NaN na primeira linha. Esse valor faltando é esperado, uma vez que não faz sentido calcular uma variação percentual para o primeiro valor de uma série. Obviamente não existe um valor anterior ao primeiro.

In [23]:
dados_ibov.return_close = dados_ibov.close.pct_change()
dados_ibov.head()

open      high       low     close  adj_close    volume  \
date                                                                      
2023-08-03  120859.0  122619.0  120365.0  120586.0   120586.0  13027300   
2023-08-04  120586.0  121442.0  119215.0  119508.0   119508.0  15065200   
2023-08-07  119509.0  120104.0  118894.0  119380.0   119380.0   8726500   
2023-08-08  119378.0  119553.0  117492.0  119090.0   119090.0  10246000   
2023-08-09  119090.0  119090.0  117901.0  118409.0   118409.0  11245600   

            return_close  
date                      
2023-08-03           NaN  
2023-08-04     -0.008940  
2023-08-07     -0.001071  
2023-08-08     -0.002429  
2023-08-09     -0.005718

Vamos fazer a mesma operação para a coluna `adj_close`.

In [24]:
dados_ibov['return_adj_close'] = dados_ibov.adj_close.pct_change()
dados_ibov.head()

open      high       low     close  adj_close    volume  \
date                                                                      
2023-08-03  120859.0  122619.0  120365.0  120586.0   120586.0  13027300   
2023-08-04  120586.0  121442.0  119215.0  119508.0   119508.0  15065200   
2023-08-07  119509.0  120104.0  118894.0  119380.0   119380.0   8726500   
2023-08-08  119378.0  119553.0  117492.0  119090.0   119090.0  10246000   
2023-08-09  119090.0  119090.0  117901.0  118409.0   118409.0  11245600   

            return_close  return_adj_close  
date                                        
2023-08-03           NaN               NaN  
2023-08-04     -0.008940         -0.008940  
2023-08-07     -0.001071         -0.001071  
2023-08-08     -0.002429         -0.002429  
2023-08-09     -0.005718         -0.005718